### ISSUE TITLE
Check implementation consistency for an overloaded function

### ISSUE THREAD

**Issue by dhruvmanila** [astral-sh/ty#109]
The task here is to check whether the overloaded signatures is consistent with the implementation signature if it's present. Specifically, from the spec:

> If an overload implementation is defined, type checkers should validate that it is consistent with all of its associated overload signatures. The implementation should accept all potential sets of arguments that are accepted by the overloads and should produce all potential return types produced by the overloads. In typing terms, this means the input signature of the implementation should be [assignable](https://typing.python.org/en/latest/spec/glossary.html#term-assignable) to the input signatures of all overloads, and the return type of all overloads should be assignable to the return type of the implementation.
> 
> If the implementation is inconsistent with its overloads, a type checker should report an error.
>
> https://typing.python.org/en/latest/spec/overload.html#implementation-consistency

The following function is the one where all overload checks happen:

https://github.com/astral-sh/ruff/blob/bd5fb826bce149d8a834767e6c175c2e4d90105a/crates/red_knot_python_semantic/src/types/infer.rs#L990-L990

Specifically, it loops over each of the overloaded function that needs to be checked here:

https://github.com/astral-sh/ruff/blob/bd5fb826bce149d8a834767e6c175c2e4d90105a/crates/red_knot_python_semantic/src/types/infer.rs#L1036-L1039

which are used to perform various checks.

**Notes:**
* We can reuse the [`INVALID_OVERLOAD`](https://github.com/astral-sh/ruff/blob/bd5fb826bce149d8a834767e6c175c2e4d90105a/crates/red_knot_python_semantic/src/types/diagnostic.rs#L487-L487) lint rule to raise a diagnostic for this check

**Comment by LaBatata101** [comment]
Can I try this one?

**Comment by dhruvmanila** [comment]
Sure!

**Comment by j178** [comment]
Hi, I ran into an issue when playing with this.

When using `Signature::is_assignable_to(f1, f2)`, it's supposed to check if  the **return type** of `f1` is assignable to the return type of `f2`, and if the **input type** of `f2` is assignable to the input type `f1`.

Then the following snippet will error becase the implementation input type (`str|int`) is not assignable to `int` or `str`:

```py
from typing import overload

@overload
def f(x: int) -> int: pass
@overload
def f(x: str) -> str: pass
def f(x: str|int) -> str|int:  # error: [invalid-overload]
    return x
```

but neither `mypy` nor `pyright` do not report error on this: [mypy playground](https://mypy-play.net/?mypy=latest&python=3.12&gist=afa534e1eae815ceab3f91930c562acb), [pyright playground](https://pyright-play.net/?code=GYJw9gtgBALgngBwJYDsDmUkQWEMpgBuApiADZgCGAJgDRQDClZZlARmcQFBcACRpCjS7ViwKMAAUADwBcmFDACUUALQA%2BBTHkJKAZz18B5KtRFiJM%2BXpggVGqDZA79h0eKlzHtgD6plappOforyUADEUChgUKTgIFxQSVAgxDAAriAoUNJcQA)

I also tried this code in `pyright`: [pyright playground](https://pyright-play.net/?code=GYJw9gtgBALgngBwJYDsDmUkQWEMpgBuApiADZgCGAJgFC0ACRpFNt1xwUwAFAB4AuTChgBKKAFoAfMJhCElAM6LGzclTocuvQVEUwQ46XoPylKrd35CARnBjFFRmXYeKBtKF6ghiMAK4gKFB8tEA)

```py
from typing import overload

@overload
def f(x: int) -> int: pass
@overload
def f(x: str) -> str: pass
def f(x: bytes) -> bytes:
    return x
```

Here’s what pyright says:
```
Overloaded implementation is not consistent with signature of overload 1
  Type "(x: bytes) -> bytes" is not assignable to type "(x: int) -> int"
    Parameter 1: type "int" is incompatible with type "bytes"
      "int" is not assignable to "bytes"
    Function return type "int" is incompatible with type "bytes"
      "int" is not assignable to "bytes"  (reportInconsistentOverload)
```

Looks like that pyright and mypy are checking both input type AND return type of each overload are assignable to the corresponding types in the implementation?

**Comment by carljm** [comment]
@j178 Yes, signature assignability is not actually the right operation here, the sense is reversed for parameters vs return type. We sorted this out in https://github.com/astral-sh/ruff/pull/17800 but I guess we didn't update the issue description here.

Note also that @LaBatata101 already has a good PR up for this in https://github.com/astral-sh/ruff/pull/17800 -- it's just waiting on a different fix so it doesn't add too many false positives.

**Comment by dhruvmanila** [comment]
(https://github.com/astral-sh/ty/issues/3118 has some additional information and examples.)

**Comment by carljm** [comment]
We seem to fully pass the typing conformance suite for this now; I think some PRs for this landed without referencing this issue?

Closing this -- feel free to reopen (or maybe just file a new issue) if there's something important we are missing here still.

**Comment by AlexWaygood** [comment]
We currently skip the check entirely for generic functions, which the conformance suite conveniently doesn't test. They proved much more complicated here than non-generic functions, but other type checkers do better than us here. @charliermarsh has looked at it a bit. We should open a followup issue.

**Comment by charliermarsh** [comment]
Good call: https://github.com/astral-sh/ty/issues/3511. I have a draft PR [here](https://github.com/astral-sh/ruff/pull/24937) but it's >2,000 lines.

**Comment by harsh-at-ikigailabs** [comment]
maybe a related issue that'd get resolved alongside this:

```python

if TYPE_CHECKING:
    import Client
    import AsyncClient

...

@typing.overload
@classmethod
def _upsert(cls, client: type[Client], **kwargs) -> str: ...  # error 1!

@typing.overload
@classmethod
def _upsert(cls, client: AsyncClient, **kwargs) -> Awaitable[str]: ...

@classmethod
def _upsert(cls, client: type[Client] | AsyncClient, **kwargs) -> str | Awaitable[str]:  # error 1!
    ...
```

With it being used like:

```python
async def foo(self, client: AsyncClient) -> str:
    val = await self._upsert(client=client, ...)  # error 2!
    return val
```

error 1: Invalid subscript of object of type `None` in a parameter annotation

error 2: `str` is not awaitable info: `__await__` is missing

 ---

If you think this is something separate then I can file a new issue.

Related: #2278 #3250 

---

If I swap `type` to `typing.Type` then both errors go away. I am using python 3.10, so using `type` instead of `typing.Type` should be identical but that's not the case here.

My ty.toml configuration:
```toml
[environment]
python = ".venv"
python-version = "3.10"
```

**Comment by dhruvmanila** [comment]
@harsh-at-ikigailabs can you provide a complete example for me to try it out? I'm not sure what `import AsyncClient` or `import Client` should resolve to?
